lib/spack/spack/spec.py

```diff
@@ -4166,134 +4288,62 @@ def splice(self, other, transitive):
-        else:
-            # NOTE: Does not fully validate providers; loader races possible
-            other_nodes = dict(
-                (s.name, s.copy(deps=False))
-                for s in other.traverse(root=True)
-                if s is other or s.name not in self
-            )
-
-        nodes = other_nodes.copy()
-        nodes.update(self_nodes)
-
-        for name in nodes:
-            if name in self_nodes:
-                for edge in self[name].edges_to_dependencies():
-                    dep_name = deps_to_replace.get(edge.spec, edge.spec).name
-                    nodes[name].add_dependency_edge(
-                        nodes[dep_name], depflag=edge.depflag, virtuals=edge.virtuals
-                    )
-                if any(dep not in self_nodes for dep in self[name]._dependencies):
-                    nodes[name].build_spec = self[name].build_spec
-            else:
-                for edge in other[name].edges_to_dependencies():
-                    nodes[name].add_dependency_edge(
-                        nodes[edge.spec.name], depflag=edge.depflag, virtuals=edge.virtuals
-                    )
-                if any(dep not in other_nodes for dep in other[name]._dependencies):
-                    nodes[name].build_spec = other[name].build_spec
-
-        ret = nodes[self.name]
-
-        # Clear cached hashes for all affected nodes
-        # Do not touch unaffected nodes
-        for dep in ret.traverse(root=True, order="post"):
-            opposite = other_nodes if dep.name in self_nodes else self_nodes
-            if any(name in dep for name in opposite.keys()):
-                # package hash cannot be affected by splice
-                dep.clear_cached_hashes(ignore=["package_hash"])
-
-                dep.dag_hash()
+        if not any(
+            node._splice_match(other, self_root=self, other_root=other)
+            for node in self.traverse(root=False, deptype=dt.LINK | dt.RUN)
+        ):
+            other_str = other.format("{name}/{hash:7}")
+            self_str = self.format("{name}/{hash:7}")
+            msg = f"Cannot splice {other_str} into {self_str}."
+            msg += f" Either {self_str} cannot depend on {other_str},"
+            msg += f" or {other_str} fails to provide a virtual used in {self_str}"
+            raise SpliceError(msg)
+
+        # Copies of all non-build deps, build deps will get added at the end
+        spec = self.copy(deps=dt.ALL & ~dt.BUILD)
+        replacement = other.copy(deps=dt.ALL & ~dt.BUILD)
+
+        # For an intransitive splice, we do a reverse transitive splice
+        # to handle the shared dependencies, and then a single pass to
+        # handle the intransitive splice itself
+        base = spec if transitive else replacement
+        source = self if transitive else other
+        new = replacement if transitive else spec
+        new_source = other if transitive else self
```

Comment:
I got lost enough in the terms here (self, other, spec, replacement, base, source, new, new_source) that I think it is better to be a little redundant than to introduce 4 new names for the reader to keep in their head.  It was a lot of cognitive load, at least for me.

This is all for setting up 3 lines of code so something like this I think makes it much more clear what is going on:

```python
def zero_build_deps(spec):
    for edge in spec.traverse_edges(cover="edges"):
        edge.depflag &= ~dt.BUILD

def make_pairs(left, right, deptype):
    return list(zip(left.traverse(deptype=deptype), right.traverse(deptype=deptype)))

if transitive:
    node_pairs = make_pairs(self, spec, dt.ALL & ~dt.BUILD)
    zero_build_deps(spec)
    spec._splice_helper(replacement, self_root=self, other_root=other)
else:
    node_pairs = make_pairs(other, replacement, dt.ALL & ~dt.BUILD)
    zero_build_deps(replacement)
    replacement._splice_helper(spec, self_root=other, other_root=self)
```

Reply:
We can't use nested methods since specs need to be pickleable -- I'll just write out the whole thing for now and we can evaluate from there.

Reply:
I think it’s just the nested method *object* that isn’t pickleable, but this wouldn’t store a nested method as a variable on the spec. Spec already has a number of methods with nested methods — format(), others.